File: MiniDOMOffice/Office/GDE/Templates/TemplateItem.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica

Imports System.Drawing
Imports minidom.XML

Partial Class Office

    Public Enum TemplateItemTypes
        NoOperation = 0
        TEXTOUT = 1
        DRAWRECT = 2
        FILLRECT = 3
        DRAWELLIPSE = 4
        FILLELLIPSE = 5
        DRAWIMAGE = 6
        NEWPAGE = 7
        DATAFIELD = 8
        EXPRESSION = 9
        GOTOPAGE = 10
    End Enum

    <Serializable>
    Public Class TemplateItem
        Inherits DBObjectBase

        Private m_ItemType As TemplateItemTypes
        Private m_Color As String
        Private m_Bounds As CRectangle
        Private m_Text As String
        Private m_Parameters As CKeyCollection

        Public Sub New()
            Me.m_ItemType = TemplateItemTypes.NoOperation
            Me.m_Color = ""
            Me.m_Bounds = New CRectangle(0, 0, 0, 0)
            Me.m_Text = ""
            Me.m_Parameters = Nothing
        End Sub

        Public Sub New(ByVal type As TemplateItemTypes, ByVal text As String, ByVal bounds As CRectangle)
            Me.New()
            Me.m_ItemType = type
            Me.m_Text = text
            Me.m_Bounds = bounds
            Me.m_Color = "#000000"
        End Sub

        Public Sub New(ByVal type As TemplateItemTypes, ByVal text As String, ByVal bounds As CRectangle, ByVal color As String)
            Me.New()
            Me.m_ItemType = type
            Me.m_Text = text
            Me.m_Bounds = bounds
            Me.m_Color = color
        End Sub

        Public ReadOnly Property Parameters As CKeyCollection
            Get
                If (Me.m_Parameters Is Nothing) Then Me.m_Parameters = New CKeyCollection
                Return Me.m_Parameters
            End Get
        End Property

        Public Property ItemType As TemplateItemTypes
            Get
                Return Me.m_ItemType
            End Get
            Set(value As TemplateItemTypes)
                Dim oldValue As TemplateItemTypes = Me.m_ItemType
                If (oldValue = value) Then Exit Property
                Me.m_ItemType = value
                Me.DoChanged("ItemType", value, oldValue)
            End Set
        End Property

        Public Property Color As String
            Get
                Return Me.m_Color
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_Color
                value = Strings.Trim(value)
                If (oldValue.Equals(value)) Then Exit Property
                Me.m_Color = value
                Me.DoChanged("Color", value, oldValue)
            End Set
        End Property

        Public Property Bounds As CRectangle
            Get
                Return Me.m_Bounds
            End Get
            Set(value As CRectangle)
                Dim oldValue As CRectangle = Me.m_Bounds
                If (Object.ReferenceEquals(oldValue, value)) Then Exit Property
                Me.m_Bounds = value
                Me.DoChanged("Bounds", value, oldValue)
            End Set
        End Property

        Public Property Text As String
            Get
                Return Me.m_Text
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_Text
                If (oldValue = value) Then Exit Property
                Me.m_Text = value
                Me.DoChanged("Text", value, oldValue)
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return [Enum].GetName(GetType(TemplateItemTypes), Me.m_ItemType) & "/" & Me.m_Bounds.ToString & "/" & Me.m_Text
        End Function

        Protected Overrides Function GetConnection() As CDBConnection
            Return Office.Database
        End Function

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_DocumentiTemplateItems"
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_ItemType = reader.Read("ItemType", Me.m_ItemType)
            Me.m_Color = reader.Read("Color", Me.m_Color)
            Dim x, y, w, h As Single
            x = reader.Read("x", x)
            y = reader.Read("y", y)
            w = reader.Read("w", w)
            h = reader.Read("h", h)
            Me.m_Bounds = New CRectangle(x, y, w, h)
            Me.m_Text = reader.Read("Text", Me.m_Text)
            Dim txt As String = reader.Read("Params", "")
            If (txt <> "") Then Me.m_Parameters = XML.Utils.Serializer.Deserialize(txt)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("ItemType", Me.m_ItemType)
            writer.Write("Color", Me.m_Color)
            writer.Write("x", Me.Bounds.Left)
            writer.Write("y", Me.Bounds.Top)
            writer.Write("w", Me.Bounds.Width)
            writer.Write("h", Me.Bounds.Height)
            writer.Write("Text", Me.m_Text)
            writer.Write("Params", XML.Utils.Serializer.Serialize(Me.Parameters))
            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Sub XMLSerialize(writer As XMLWriter)
            writer.WriteAttribute("ItemType", Me.m_ItemType)
            writer.WriteAttribute("Color", Me.m_Color)
            MyBase.XMLSerialize(writer)
            writer.WriteTag("Bounds", Me.Bounds)
            writer.WriteTag("Params", Me.Parameters)
            writer.WriteTag("Text", Me.m_Text)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "ItemType" : Me.m_ItemType = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Color" : Me.m_Color = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Bounds" : Me.m_Bounds = XML.Utils.Serializer.ToObject(fieldValue)
                Case "Text" : Me.m_Text = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Params" : Me.m_Parameters = fieldValue
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select

        End Sub

    End Class


End Class

File: MiniDOMOffice/Office/GDE/Templates/Items/TemplateItem.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;

namespace minidom
{
    public partial class Office
    {
        /// <summary>
        /// Tipi di elementi supportati in un template
        /// </summary>
        public enum TemplateItemTypes
        {
            /// <summary>
            /// L'elemento non é renderizzabile
            /// </summary>
            NoOperation = 0,

            /// <summary>
            /// Etichetta
            /// </summary>
            TEXTOUT = 1,

            /// <summary>
            /// Bordi del Rettangolo
            /// </summary>
            DRAWRECT = 2,

            /// <summary>
            /// Riempimento del rettangolo
            /// </summary>
            FILLRECT = 3,

            /// <summary>
            /// Bordo dell'ellisse
            /// </summary>
            DRAWELLIPSE = 4,

            /// <summary>
            /// Rimempimento dell'ellisse
            /// </summary>
            FILLELLIPSE = 5,

            /// <summary>
            /// Immagine
            /// </summary>
            DRAWIMAGE = 6,

            /// <summary>
            /// Genera una nuova pagina
            /// </summary>
            NEWPAGE = 7,

            /// <summary>
            /// Campo data
            /// </summary>
            DATAFIELD = 8,

            /// <summary>
            /// Campo espressione
            /// </summary>
            EXPRESSION = 9,

            /// <summary>
            /// Vai alla pagina
            /// </summary>
            GOTOPAGE = 10
        }


        /// <summary>
        /// Elemento generato da un template
        /// </summary>
        [Serializable]
        public class TemplateItem 
            : minidom.Databases.DBObjectBase
        {
            private TemplateItemTypes m_ItemType;
            private string m_Color;
            private CRectangle m_Bounds;
            private string m_Text;
            
            /// <summary>
            /// Costruttore
            /// </summary>
            public TemplateItem()
            {
                m_ItemType = TemplateItemTypes.NoOperation;
                m_Color = "";
                m_Bounds = new CRectangle(0d, 0d, 0d, 0d);
                m_Text = "";
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="type"></param>
            /// <param name="text"></param>
            /// <param name="bounds"></param>
            public TemplateItem(TemplateItemTypes type, string text, CRectangle bounds) : this()
            {
                m_ItemType = type;
                m_Text = text;
                m_Bounds = bounds;
                m_Color = "#000000";
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="type"></param>
            /// <param name="text"></param>
            /// <param name="bounds"></param>
            /// <param name="color"></param>
            public TemplateItem(TemplateItemTypes type, string text, CRectangle bounds, string color) : this()
            {
                m_ItemType = type;
                m_Text = text;
                m_Bounds = bounds;
                m_Color = color;
            }
             
            /// <summary>
            /// Tipo dell'elemento
            /// </summary>
            public TemplateItemTypes ItemType
            {
                get
                {
                    return m_ItemType;
                }

                set
                {
                    var oldValue = m_ItemType;
                    if (oldValue == value)
                        return;
                    m_ItemType = value;
                    DoChanged("ItemType", value, oldValue);
                }
            }

            /// <summary>
            /// Colore
            /// </summary>
            public string Color
            {
                get
                {
                    return m_Color;
                }

                set
                {
                    string oldValue = m_Color;
                    value = DMD.Strings.Trim(value);
                    if (oldValue.Equals(value))
                        return;
                    m_Color = value;
                    DoChanged("Color", value, oldValue);
                }
            }

            /// <summary>
            /// Rettangolo che contiene l'elemento
            /// </summary>
            public CRectangle Bounds
            {
                get
                {
                    return m_Bounds;
                }

                set
                {
                    var oldValue = m_Bounds;
                    if (ReferenceEquals(oldValue, value))
                        return;
                    m_Bounds = value;
                    DoChanged("Bounds", value, oldValue);
                }
            }

            /// <summary>
            /// Testo
            /// </summary>
            public string Text
            {
                get
                {
                    return m_Text;
                }

                set
                {
                    string oldValue = m_Text;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Text = value;
                    DoChanged("Text", value, oldValue);
                }
            }

            /// <summary>
            /// Estringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return DMD.Strings.ConcatArray(
                        Enum.GetName(typeof(TemplateItemTypes), m_ItemType) , "/" ,
                        m_Bounds.ToString() , "/" ,
                        m_Text
                        );
            }

            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return null;
            }

            public override string GetTableName()
            {
                return "tbl_DocumentiTemplateItems";
            }

            protected override bool LoadFromRecordset(DBReader reader)
            {
                this.m_ItemType = reader.Read("ItemType", this.m_ItemType);
                this.m_Color = reader.Read("Color", this.m_Color);
                float x = default, y = default, w = default, h = default;
                x = reader.Read("x", x);
                y = reader.Read("y", y);
                w = reader.Read("w", w);
                h = reader.Read("h", h);
                m_Bounds = new CRectangle(x, y, w, h);
                this.m_Text = reader.Read("Text", this.m_Text);
                string txt = reader.Read("Params", "");
                if (!string.IsNullOrEmpty(txt))
                    this.m_Parameters = (CKeyCollection)DMD.XML.Utils.Serializer.Deserialize(txt);

                return base.LoadFromRecordset(reader);
            }

            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("ItemType", m_ItemType);
                writer.Write("Color", m_Color);
                writer.Write("x", Bounds.Left);
                writer.Write("y", Bounds.Top);
                writer.Write("w", Bounds.Width);
                writer.Write("h", Bounds.Height);
                writer.Write("Text", m_Text);
                writer.Write("Params", DMD.XML.Utils.Serializer.Serialize(Parameters));
                return base.SaveToRecordset(writer);
            }

            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("ItemType", (int?)m_ItemType);
                writer.WriteAttribute("Color", m_Color);
                base.XMLSerialize(writer);
                writer.WriteTag("Bounds", Bounds);
                writer.WriteTag("Params", Parameters);
                writer.WriteTag("Text", m_Text);
            }

            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "ItemType":
                        {
                            m_ItemType = (TemplateItemTypes)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Color":
                        {
                            m_Color = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Bounds":
                        {
                            m_Bounds = (CRectangle)DMD.XML.Utils.Serializer.ToObject(fieldValue);
                            break;
                        }

                    case "Text":
                        {
                            m_Text = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Params":
                        {
                            m_Parameters = (CKeyCollection)fieldValue;
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }
        }
    }
}